helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -290,7 +325,82 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
     } // end of for-each-partition
   }
 
-  private boolean shouldCreateSTCancellation(Message pendingMessage, String desiredState,
+  /**
+   * Calculate the current active replica count based on state model type.
+   * The count includes replicas in top states, secondary top states (where applicable),
+   * and ERROR states since helix considers them active.
+   * State model handling:
+   * - Single-top state models: Differentiates between patterns with and without secondary top
+   * states
+   * - ONLINE-OFFLINE: Counts ONLINE + ERROR states only
+   * - MASTER-SLAVE-OFFLINE: Counts MASTER + SLAVE + ERROR states
+   * - ONLINE-STANDBY-OFFLINE: Counts ONLINE + STANDBY + ERROR states
+   * - Multi-top state models: Counts only top states + ERROR states
```

Comment:
In multiple top state, with ONLINE-STANDBY-OFFLINE state model, dont we also consider standby as active?